pkg/admission/cert/writer/secret.go

```diff
@@ -0,0 +1,246 @@
+
+func (s *secretReadWriter) write(webhookName string) (
+	*generator.Artifacts, error) {
+	v := s.webhookMap[webhookName]
+
+	webhook := v.webhook
+	commonName, err := dnsNameForWebhook(&webhook.ClientConfig)
+	if err != nil {
+		return nil, err
+	}
+	certs, err := s.certGenerator.Generate(commonName)
+	if err != nil {
+		return nil, err
+	}
+	secret := certsToSecret(certs, v.secret)
+	// TODO: fix it: see the TODO in the method
+	err = setOwnerRef(secret, s.webhookConfig)
+	if err != nil {
+		return nil, err
+	}
+	err = s.client.Create(nil, secret)
+	return certs, err
+}
+
+func (s *secretReadWriter) overwrite(webhookName string) (
+	*generator.Artifacts, error) {
+	v := s.webhookMap[webhookName]
+
+	webhook := v.webhook
+	commonName, err := dnsNameForWebhook(&webhook.ClientConfig)
+	if err != nil {
+		return nil, err
+	}
+	certs, err := s.certGenerator.Generate(commonName)
+	if err != nil {
+		return nil, err
+	}
+	secret := certsToSecret(certs, v.secret)
+	// TODO: fix it: see the TODO in the method
+	err = setOwnerRef(secret, s.webhookConfig)
+	if err != nil {
+		return nil, err
+	}
+	err = s.client.Update(nil, secret)
+	return certs, err
+}
+
+func (s *secretReadWriter) read(webhookName string) (*generator.Artifacts, error) {
+	v := s.webhookMap[webhookName]
+	secret := &corev1.Secret{}
+	err := s.client.Get(nil, v.secret, secret)
+	return secretToCerts(secret), err
+}
+
+// setOwnerRef marks the webhook as the owner of the secret by setting the ownerReference in the secret.
+func setOwnerRef(secret, webhookConfig runtime.Object) error {
+	accessor, err := meta.Accessor(webhookConfig)
+	if err != nil {
+		return err
+	}
+	// TODO: typeAccessor.GetAPIVersion() and typeAccessor.GetKind() returns empty apiVersion and Kind, fix it.
+	typeAccessor, err := meta.TypeAccessor(webhookConfig)
+	if err != nil {
+		return err
+	}
+	blockOwnerDeletion := false
+	// Due to
+	// https://github.com/kubernetes/kubernetes/blob/5da925ad4fd070e687dc5255c177d5e7d542edd7/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/controller_ref.go#L35
+	isController := true
+	ownerRef := metav1.OwnerReference{
+		APIVersion:         typeAccessor.GetAPIVersion(),
+		Kind:               typeAccessor.GetKind(),
+		Name:               accessor.GetName(),
+		UID:                accessor.GetUID(),
+		BlockOwnerDeletion: &blockOwnerDeletion,
+		Controller:         &isController,
+	}
+	secretAccessor, err := meta.Accessor(secret)
+	if err != nil {
+		return err
+	}
+	secretAccessor.SetOwnerReferences([]metav1.OwnerReference{ownerRef})
+	return nil
+}
+
+func secretToCerts(secret *corev1.Secret) *generator.Artifacts {
+	if secret.Data == nil {
+		return nil
+	}
+	return &generator.Artifacts{
+		CACert: secret.Data[CACertName],
+		Cert:   secret.Data[ServerCertName],
+		Key:    secret.Data[ServerKeyName],
+	}
+}
+
+func certsToSecret(certs *generator.Artifacts, sec apitypes.NamespacedName) *corev1.Secret {
+	return &corev1.Secret{
+		TypeMeta: metav1.TypeMeta{
```

Comment:
FYI: you shouldn't need to fill in the TypeMeta.

Reply:
Why?
Does it mean the client.Client will set the `TypeMeta` fields for me if I pass it a `*corev1.Secret`?